File: src/test/java/unittests/AbstractTest.java
package unittests;

import core.environment.DefaultEnvironment;
import core.environment.Environment;
import core.evaluator.Evaluator;
import core.reader.StringReader;
import core.scm.SCMSymbol;

import java.util.Arrays;

import static org.junit.Assert.assertEquals;

public abstract class AbstractTest {

  private final StringReader reader = new StringReader();
  protected final Evaluator eval = new Evaluator();
  protected final DefaultEnvironment env = new DefaultEnvironment();
  {
    /* Eval lib procedures */
    for (String proc : env.getLibraryProcedures()) {
      for (Object p : reader.read(proc)) {
        eval.macroexpandAndEvaluate(p, env);
      }
    }
  }
  /* Helper method: evaluates first S-expression */
  protected Object eval(String sexp, Environment env) {
    return eval.macroexpandAndEvaluate(reader.readFirst(sexp), env);
  }

  protected SCMSymbol s(String str) {
    return SCMSymbol.of(str);
  }

  protected void assertAllEqual(Object expected, String[] forms, Environment env) {
    Arrays.stream(forms).forEach(form -> assertEquals(form, expected, eval(form, env)));
  }
}

File: src/test/java/unittests/AbstractTest.kt
package unittests

import core.environment.DefaultEnvironment
import core.Evaluator
import core.reader.StringReader
import core.scm.Symbol
import org.junit.Assert.assertEquals
import java.util.*

abstract class AbstractTest {

    private val reader = StringReader()
    private val environment = DefaultEnvironment()
    private val evaluator = Evaluator(environment).apply {
        with (reader) {
            environment.libraryProcedures.forEach { this@apply.eval(readOne(it)) }
        }
    }

    /* Helper method: evaluates first S-expression */
    protected fun eval(sexp: String) = evaluator.macroexpandAndEvaluate(reader.readOne(sexp)!!)

    protected fun s(str: String) = Symbol.intern(str)

    protected fun assertAllEqual(expected: Any, forms: Array<String>) {
        Arrays.stream(forms).forEach { assertEquals(it, expected, eval(it)) }
    }
}
